browser-sign-in/OktaBrowserSignIn/WelcomeViewController.swift

```diff
@@ -15,70 +15,88 @@
+        if let clientId = auth?.signInFlow.client.configuration.clientId {
+            self.clientIdLabel.text = "clientId: \(clientId)"
+        } else {
+            self.clientIdLabel.text = "Not configured"
+            self.signInButton.isEnabled = false
         }
     }
-
+    
+    func navigateToDetailsPage() {
+        guard Credential.default != nil else { return }
+        self.performSegue(withIdentifier: "show-details", sender: self)
+    }
+    
+    @IBAction func ephemeralSwitchChanged(_ sender: Any) {
+        guard let sender = sender as? UISwitch else { return }
+        self.auth?.ephemeralSession = sender.isOn
+    }
+    
     @IBAction private func signInTapped() {
-        if #available(iOS 13.0, *) {
-            oktaOidc?.configuration.noSSO = ephemeralSwitch.isOn
-        }
-        
-        oktaOidc?.signInWithBrowser(from: self, callback: { [weak self] stateManager, error in
-            if let error = error {
-                let alert = UIAlertController(title: "Error", message: error.localizedDescription, preferredStyle: .alert)
+        let window = viewIfLoaded?.window
+        auth?.signIn(from: window) { result in
+            switch result {
+            case .success(let token):
+                do {
+                    try Credential.store(token)
+                    self.performSegue(withIdentifier: "show-details", sender: self)
+                } catch {
+                    self.show(error: error)
+                    return
+                }
+                
+            case .failure(let error):
+                let alert = UIAlertController(title: "Error",
+                                              message: error.localizedDescription,
+                                              preferredStyle: .alert)
                 alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
-                self?.present(alert, animated: true, completion: nil)
+                self.present(alert, animated: true, completion: nil)
                 return
             }
-            self?.stateManager?.clear()
-            self?.stateManager = stateManager
-            self?.stateManager?.writeToSecureStorage()
-            self?.performSegue(withIdentifier: "show-details", sender: self)
-        })
+        }
     }
-
-    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
-        if let destinationViewController = segue.destination as? SignInViewController {
-            destinationViewController.oktaOidc = self.oktaOidc
-            destinationViewController.stateManager = self.stateManager
+    
+    func show(error: Error) {
+        // There's currently no way to know when the ASWebAuthenticationSession will be dismissed,
+        // so to ensure the alert can be displayed, we must delay presenting an error until the
+        // dismissal is complete.
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.33) {
```

Comment:
To make this generic, you can  make the delay configurable, so this one method can be used for the various error results you get.
```suggestion
    func show(error: Error, after delay: TimeInterval = 0.0) {
        // There's currently no way to know when the ASWebAuthenticationSession will be dismissed,
        // so to ensure the alert can be displayed, we must delay presenting an error until the
        // dismissal is complete.
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.33) {
```